06_puppy/nicholascross/main.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+type ID uint64
+type Breed string
+type Colour string
+type Value float64
+
+type Puppy struct {
+	ID     ID
+	Breed  Breed
+	Colour Colour
+	Value  Value
+}
+
+type Storer interface {
+	CreatePuppy(p Puppy)
+	RetrievePuppy(ID) *Puppy
+	UpdatePuppy(ID, Puppy)
+	DeletePuppy(ID)
+}
+
+type MapStore map[ID]*Puppy
+
+var out io.Writer = os.Stdout
+
+func main() {
+	p := Puppy{1, "Jack Russel Terrier", "white and brown", 550.0}
+	fmt.Fprintf(out, "%d - %s [%s]: %f", p.ID, p.Breed, p.Colour, p.Value)
+}
+
+func (ms MapStore) CreatePuppy(p Puppy) {
+	pexist := ms[p.ID]
+	if pexist == nil {
+		ms[p.ID] = &p
+	}
+}
+
+func (ms MapStore) RetrievePuppy(id ID) *Puppy {
+	return ms[id]
+}
+
+func (ms MapStore) UpdatePuppy(id ID, p Puppy) {
```

Comment:
First argument is unused and can be removed. However, with a little bit of foresight, following on lab 10
`PUT    /api/puppy/{id}      Payload: Puppy JSON without ID`
Therefore, setting id as p.ID would be better. 😉

Reply:
Oh, I didn't notice that 😄

@kasunfdo are you suggesting that I should do something like

```
p.ID = id
```

or just remove the parameter from the function?

```
func (ms MapStore) UpdatePuppy(p Puppy) 
```

Reply:
Yes. Better using a puppy instance without id as the second arg, setting `p.ID = id` if id is not in store. 
It will be beneficial for the future.